tests/process.md

```diff
@@ -0,0 +1,160 @@
+# Setting up the environment
+
+```ocaml
+# #require "eio_main";;
+```
+
+Creating some useful helper functions
+
+```ocaml
+open Eio
+open Eio.Std
+
+let spawn ~env ~sw ?cwd cmd args =
+  Process.spawn ~sw ?cwd ~stdout:env#stdout ~stdin:env#stdin ~stderr:env#stderr env#process_mgr cmd args
+
+let run fn =
+  Eio_main.run @@ fun env ->
+  fn (spawn ~env) env
+```
+
+Running a program as a subprocess
+
+```ocaml
+# run @@ fun spawn env ->
+  Switch.run @@ fun sw ->
+  let t = spawn ~sw "/usr/bin/echo" [ "echo"; "hello world" ] in
+  Process.status t;;
+hello world
+- : Process.status = Eio.Process.Exited 0
+```
+
+Stopping a subprocess works and checking the status waits and reports correctly
+
+```ocaml
+# run @@ fun spawn _env ->
+  Switch.run @@ fun sw ->
+  let t = spawn ~sw "/usr/bin/sleep" [ "sleep"; "10" ] in
+  Process.signal t Sys.sigkill;
+  Process.status t;;
+- : Process.status = Eio.Process.Signaled (-7)
+```
+
+A switch will stop a process when it is released.
+<!-- Need a better test of this... -->
+
+```ocaml
+# run @@ fun spawn env ->
+  let proc = ref None in 
+  let run () =
+    Switch.run @@ fun sw ->
+    proc := Some (spawn ~sw "/usr/bin/sleep" [ "sleep"; "10" ])
+  in
+  run ();
+  Process.status (Option.get !proc);;
+- : Process.status = Eio.Process.Signaled (-7)
+```
+
+Passing in flows allows you to redirect the child process' stdout.
+
+```ocaml
+# run @@ fun _spawn env ->
+  let process = Eio.Stdenv.process_mgr env in
+  let fs = Eio.Stdenv.fs env in
+  let filename = "process-test.txt" in
+  let run () =
+    Eio.Path.(with_open_out ~create:(`Exclusive 0o600) (fs / filename)) @@ fun stdout ->
+    let stdout = (stdout :> Eio.Flow.sink) in
+    Switch.run @@ fun sw ->
+    let t = Eio.Process.spawn ~sw ~stdout ~stdin:env#stdin ~stderr:env#stderr process "/usr/bin/echo" [ "echo"; "Hello" ] in
+    Process.status t
+  in
+  match run () with
+    | Exited 0 ->
+      Eio.Path.(with_open_in (fs / filename)) @@ fun flow ->
+      let buff = Buffer.create 128 in
+      Eio.Flow.copy flow (Eio.Flow.buffer_sink buff);
+      Buffer.contents buff
+    | _ -> failwith "Subprocess didn't exit cleanly!";;
```

Comment:
We could probably do with a helper function in lib_eio for checking the exit status. `Process.await_exn` or something?

Reply:
Good idea -- do you think instead of the `val exit_status` we should provide `val await : t -> exit_status` and `val await_exn : t -> unit` ?